crates/proc_macro_api/src/lib.rs

```diff
@@ -108,4 +122,71 @@ impl ProcMacroClient {
             }
         }
     }
+
+    // This is used inside self.read_version() to locate the ".rustc" section
+    // from a proc macro crate's binary file.
+    fn read_section<'a>(
+        &'a self,
+        dylib_binary: &'a [u8],
+        section_name: &'a str,
+    ) -> io::Result<&'a [u8]> {
+        BinaryFile::parse(dylib_binary)
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?
+            .section_by_name(section_name)
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "section read error"))?
+            .data()
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+    }
+
+    // Check the version of rustc that was used to compile a proc macro crate's
+    // binary file.
+    // A proc macro crate binary's ".rustc" section has following byte layout:
+    // * [b'r',b'u',b's',b't',0,0,0,5] is the first 8 bytes
+    // * ff060000 734e6150 is followed, it's the snappy format magic bytes,
+    //   means bytes from here(including this sequence) are compressed in
+    //   snappy compression format. Version info is here inside, so decompress
+    //   this.
+    // The bytes you get after decompressing the snappy format portion has
+    // following layout:
+    // * [b'r',b'u',b's',b't',0,0,0,5] is the first 8 bytes(again)
+    // * [crate root bytes] next 4 bytes is to store crate root position,
+    //   according to rustc's source code comment
+    // * [length byte] next 1 byte tells us how many bytes we should read next
+    //   for the version string's utf8 bytes
+    // * [version string bytes encoded in utf8] <- GET THIS BOI
+    // * [some more bytes that we don really care but still there] :-)
+    // Check this issue for more about the bytes layout:
+    // https://github.com/rust-analyzer/rust-analyzer/issues/6174
+    fn read_version(&self, dylib_path: &Path) -> io::Result<String> {
```

Comment:
Is this actually getting called somewhere?

Reply:
Not yet, I think. We need to call it before loading the library, around `ProcMacroLibraryLibloading::open` or above.

But we still need to figure out what version to expect.

Reply:
No, but there is this [discussion](https://github.com/rust-analyzer/rust-analyzer/issues/6174#issuecomment-743150971).

Reply:
Well, it would still be nice to wire it up. I just didn't want to scare you off right from the start 😬

Reply:
> Well, it would still be nice to wire it up. I just didn't want to scare you off right from the start grimacing

I see, sure thing. :smile:

Reply:
> Not yet, I think. We need to call it before loading the library, around `ProcMacroLibraryLibloading::open` or above.
> 
> But we still need to figure out what version to expect.

I guess I should do version check before `let lib = load_library(file).map_err(invalid_data_err)?;`? In that case, I better move this method under `ProcMacroLibraryLibloading` type? (:sweat_smile: I thought I should do version check inside `ProcMacroClient::by_dylib_path`, but looks like library loading is actually happening in the method you mentioned, so moving in here seems more resonable)

Reply:
> So I think the last ABI change was [rust-lang/rust#74653](https://github.com/rust-lang/rust/pull/74653), which means the earliest `rustc` we support is... the nightly from `2020-07-28`?
> `rustc 1.47.0-nightly (6c8927b0c 2020-07-26)` doesn't work
> `rustc 1.47.0-nightly (76e83339b 2020-07-27)` works

So the check logic should be like, the binary's version string's date should be no earlier than `2020-07-27`?

Reply:
> So the check logic should be like, the binary's version string's date should be no earlier than 2020-07-27?

Yes, I guess you can take the date from the version string and compare it (as a string) with `2020-07-27`.

Reply:
> > Not yet, I think. We need to call it before loading the library, around `ProcMacroLibraryLibloading::open` or above.
> > But we still need to figure out what version to expect.
> 
> I guess I should do version check before `let lib = load_library(file).map_err(invalid_data_err)?;`? (actually, I think I should check version even before checking symbol?)In that case, I better move this method under `ProcMacroLibraryLibloading` type? (sweat_smile I thought I should do version check inside `ProcMacroClient::by_dylib_path`, but looks like library loading is actually happening in the method you mentioned, so moving in here seems more resonable)

And this, moving the method under `ProcMacroLibraryLibloading`? Now it's under `ProcMacroClient`.

Reply:
That won't quite work for beta and stable, as those don't show the date at which the branch forked from master (and thus abi compatibility) but the date at which they were built.

Reply:
Oh, right. We need three version strings, then?

I hate this.

Reply:
(confused and scratching my head)

Reply:
So how do I get the version that I am comparing with for stable and beta? And is there universal way of doing this for all three including nightly?

Reply:
We have some discussion in zulip that only stable is fine for now.